application/test/controllers/widgets_controller_test.rb

```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+require 'test_helper'
+
+class WidgetsControllerTest < ActionDispatch::IntegrationTest
+  test 'should return 400 for invalid widget path' do
+    get widgets_path(widget_path: '../etc/passwd')
+    assert_response :bad_request
+    assert_includes @response.body, '400 Bad Request'
+  end
```

Comment:
I wonder what would happen if I use the widget_path `$HOME/private/file` on this controller. Have you filtered this kind of inputs?